HLTrigger/Timer/plugins/ThroughputService.cc

```diff
@@ -68,19 +90,65 @@ void ThroughputService::preGlobalBeginRun(edm::GlobalContext const& gc) {
     // book MonitorElement's for this run
     edm::Service<DQMStore>()->meBookerGetter(bookTransactionCallback);
   } else {
-    std::cerr << "No DQMStore service, aborting." << std::endl;
-    abort();
+    m_sourced_events = nullptr;
+    m_retired_events = nullptr;
   }
 }
 
 void ThroughputService::preSourceEvent(edm::StreamID sid) {
-  auto timestamp = std::chrono::steady_clock::now();
-  m_sourced_events->Fill(std::chrono::duration_cast<std::chrono::duration<double>>(timestamp - m_startup).count());
+  auto timestamp = std::chrono::high_resolution_clock::now();
+  auto interval = std::chrono::duration_cast<std::chrono::duration<double>>(timestamp - m_startup).count();
+  if (m_enable_dqm) {
+    m_sourced_events->Fill(interval);
+  }
 }
 
 void ThroughputService::postEvent(edm::StreamContext const& sc) {
-  auto timestamp = std::chrono::steady_clock::now();
-  m_retired_events->Fill(std::chrono::duration_cast<std::chrono::duration<double>>(timestamp - m_startup).count());
+  auto timestamp = std::chrono::high_resolution_clock::now();
+  auto interval = std::chrono::duration_cast<std::chrono::duration<double>>(timestamp - m_startup).count();
+  if (m_enable_dqm) {
+    m_retired_events->Fill(interval);
+  }
+  ++m_counter;
+  if (m_counter % m_resolution == 0) {
+    m_events.push_back(timestamp);
+  }
+}
+
+void ThroughputService::postEndJob() {
+  if (m_counter < 2 * m_resolution) {
+    // not enough mesurements to estimate the throughput
+    edm::LogWarning("ThroughputService") << "Not enough events to measure the throughput with a resolution of "
+                                         << m_resolution << " events";
+    return;
+  }
+
+  edm::LogInfo info("ThroughputService");
+
+  if (m_print_event_summary) {
+    for (uint32_t i = 0; i < m_events.size(); ++i) {
+      info << std::setw(8) << (i + 1) * m_resolution << ", " << std::setprecision(6)
+           << edm::TimeOfDay(std::chrono::duration_cast<std::chrono::microseconds>(m_events[i].time_since_epoch()))
+           << std::setfill(' ')  // workaround for a bug (?) in operator<<(std::ostream&, TimeOfDay)
```

Comment:
For the moment, I've added this as a workaround.

Reply:
> For the moment, I've added this as a workaround.

Do I understand correctly that this fix
https://github.com/cms-sw/cmssw/blob/07ab9b0f09890bc3ad06a4953d6ec92d52c0ec35/FWCore/Utilities/src/TimeOfDay.cc#L29
made this workaround unnecessary?

Reply:
Yes - I've removed the workaround.